src/libAtomVM/ets_hashtable.c

```diff
@@ -81,9 +81,49 @@ static void print_info(struct EtsHashTable *hash_table)
     }
 }
 #endif
+struct HNode *ets_hashtable_new_node(term entry, int keypos, GlobalContext *global)
+{
+
+    Heap *heap = malloc(sizeof(Heap));
+    if (IS_NULL_PTR(heap)) {
+        return NULL;
+    }
+    size_t size = (size_t) memory_estimate_usage(entry);
+    if (memory_init_heap(heap, size) != MEMORY_GC_OK) {
+        free(heap);
+        return NULL;
+    }
+
+    term new_entry = memory_copy_term_tree(heap, entry);
+    struct HNode *new_node = malloc(sizeof(struct HNode));
+    if (IS_NULL_PTR(new_node)) {
+        memory_destroy_heap(heap, global);
+        return NULL;
+    }
+    term key = term_get_tuple_element(new_entry, keypos);
+
+    new_node->next = NULL;
+    new_node->key = key;
+    new_node->entry = new_entry;
+    new_node->heap = heap;
 
-EtsHashtableErrorCode ets_hashtable_insert(struct EtsHashTable *hash_table, term key, term entry, EtsHashtableOptions opts, Heap *heap, GlobalContext *global)
+    return new_node;
+}
+
+void free_hashtable_node_array(struct HNode **allocated, size_t size, GlobalContext *global)
```

Comment:
I'm not sure if this should live here or in `ets.c` instead.

Reply:
Maybe yes (since it is a helper specific to an implementation detail in ets.c), and maybe here in this place we just hide a bit the detail of the heap we are creating with a destroy function that just frees both the node and the heap).
Pro-tip: let's remember to use `static` to keep helper functions "private", in this case we make this a helper function